### PR TITLE
Extend `gem` DSL with a `force_ruby_platform` option

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
## What was the end-user or developer problem that led to this PR?

Sometimes it would be useful, specially when we enable the `specific_platform` setting by default to be able to opt-out of the standard platform priority and always force the ruby variant for specific gems.

## What is your fix for the problem, implemented in this PR?

Add a `:force_ruby_platform` option to the `gem` DSL.

This is the "per gem" equivalent of the `force_ruby_platform` setting.

This can be handy (assuming the pure ruby variant works just fine) when:

* You're having issues with the platform specific variant.

* The platform specific variant does not yet support a newer ruby (and thus has a `required_ruby_version` upper bound), but you still want your Gemfile{.lock} files to resolve under that ruby.

## Make sure he following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

This PR closes https://github.com/rubygems/rubygems/issues/3347.

### CODE REVIEW

File: bundler/spec/install/gemfile/force_ruby_platform_spec.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    it "still pulls the ruby variant" do
+      simulate_platform "ruby" do
+        install_gemfile <<-G
+          source "#{file_uri_for(gem_repo4)}"
+
+          gem "depends_on_platform_specific"
+          gem "platform_specific", :force_ruby_platform => true
+        G
+
+        expect(the_bundle).to include_gems "platform_specific 1.0.0 RUBY"
```
Comment (simi) [inline bundler/spec/install/gemfile/force_ruby_platform_spec.rb:60]: :thinking: It would make sense to me to add also specs for not setting `force_ruby_platform` here to see (and actually test) the difference.

```ruby
gem "platform_specific-forced", :force_ruby_platform => true
gem "platform_specific"
# ...
expect(the_bundle).to include_gems "platform_specific-forced 1.0.0 RUBY"
expect(the_bundle).to include_gems "platform_specific 1.0.0 JAVA"
```

Comment (deivid-rodriguez) [reply bundler/spec/install/gemfile/force_ruby_platform_spec.rb:60]: Sure.

**Review by simi — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
I'd like to introduce this, we probably need a new minor but it seems ok? I was kind of waiting to introduce #4076 before this, because it's a really nice feature to have before introducing DSL changes.

With that in place, you could freely use new `DSL` features in your Gemfile like `gem "google-protobuf", force_ruby_platform: true` and happily distribute them, knowing that all users of your Gemfile will transparently get a bundler version that understands those features. Of course, people would need to run a bundler version that has #4076. Maybe not a life changer, but it sounds really nice to me.

If we proceeed with this PR _without_ automatic installation of the `BUNDLED WITH` version, then users not running a proper version will get an error like:

> You passed `:force_ruby_platform` as an option for gem "google-protobuf", but it is invalid. Valid options are: <list_of_valid_options>. You may be able to resolve this by upgrading Bundler to the newest version.

Not as nice as "just works", but not that bad either.

**Comment by ncri** [comment]
When is this going to be merged?

**Comment by ncri** [comment]
Anyone reading here? Looks this has already been approved?

**Comment by ngan** [comment]
@deivid-rodriguez will this make it into next release? 🙏

**Comment by deivid-rodriguez** [comment]
It depends on whether I manage to make it actually work or not 😆,  but I hope to be able to release it soon!

**Comment by ncri** [comment]
That would be excellent! :)

**Comment by eregon** [comment]
This functionality would be great, recently @bjfish and I were wondering if `gem 'foo', platforms: :ruby` was related but actually not, that Bundler `gem name, platforms:` is completely unrelated to the the `Gem.platforms`.

File: bundler/lib/bundler/dependency.rb [hunk 2]
```diff
@@ -159,5 +159,13 @@ def specific?
     rescue NoMethodError
       requirement != ">= 0"
     end
+
+    private
+
+    def default_force_ruby_platform
+      return false unless Bundler.current_ruby.truffleruby?
+
+      !%w[libv8 sorbet-static].include?(name)
```
Comment (eregon) [inline bundler/lib/bundler/dependency.rb:168]: (reposting it here so we can discuss this on the PR more easily, vs a comment on a commit)
Could we use `match_gem?` here so we avoid duplicating the logic of https://github.com/oracle/truffleruby/blob/b131d3c146afe6bed0f119d0891b34ba68ac5deb/lib/mri/rubygems/platform.rb#L38-L47 ?
That list is subject to change and needs to remain in the truffleruby repo.

Comment (deivid-rodriguez) [reply bundler/lib/bundler/dependency.rb:168]: This was just a WIP but I was planning to use the constant.

Comment (deivid-rodriguez) [reply bundler/lib/bundler/dependency.rb:168]: Truffleruby wants to force the ruby platform for some specific gems, using this feature seems like a perfect use case for me.

Comment (eregon) [reply bundler/lib/bundler/dependency.rb:168]: > Truffleruby wants to force the ruby platform for some specific gems, using this feature seems like a perfect use case for me.

Ah right, this would be a cleaner way to integrate than to override `match_gem?` in TruffleRuby, is that what you mean?
That sounds good in general.
I'd be more comfortable if this `default_force_ruby_platform` was a method on e.g. `Gem` (and it would be passed the gem name and the version, as the answer might depend on the version) and TruffleRuby could override that method in `rubygems/defaults/truffleruby.rb`

Comment (deivid-rodriguez) [reply bundler/lib/bundler/dependency.rb:168]: > Is there are reason we can't use match_gem? here?

We can, but using this feature feels way better to me. As you can see in d3317b8192cc912f9219c84f2b6302548be1fe34, it allows us to remove all the hacks we need to keep just for truffleruby, and that we introduced because of causing regressions. I think this is much cleaner and less likely to cause issues.

Comment (deivid-rodriguez) [reply bundler/lib/bundler/dependency.rb:168]: > I'd be more comfortable if this default_force_ruby_platform was a method on e.g. Gem (and it would be passed the gem name and ideally the version too, as the answer might depend on the version) and TruffleRuby could override that method in rubygems/defaults/truffleruby.rb

I think passing the version is overkill to me. We haven't run resolution at this point yet, so we only have Gemfile information here. We could use the version if it's explicit in the Gemfile, but I think it's an unnecessary complication. Let's stick to this. Regarding moving the method, I don't think this is a good fit as a global `Gem` method, and I don't want to pollute `Gem` more unnecessarily. Why will you want to override this method?

Comment (deivid-rodriguez) [reply bundler/lib/bundler/dependency.rb]: I added a commit to use `Gem::Platform::REUSE_AS_BINARY_ON_TRUFFLERUBY`. I'm happy to backfill it for older versions of truffleruby that didn't have it.

Comment (eregon) [reply bundler/lib/bundler/dependency.rb]: Yes, agreed having the version is not worth the trouble at this point, just the gem name is enough currently.

> Regarding moving the method, I don't think this is a good fit as a global Gem method, and I don't want to pollute Gem more unnecessarily. 

It could be somewhere else, e.g., on Gem::Platform or any other place.
An instance method on Bundler::Dependency feels weird to me if we ever need that logic anywhere else, but maybe we don't, WDYT?
I guess `Bundler::Dependency#force_ruby_platform` is already a good way to get the result of this check.
An advantage of having it on Bundler::Dependency is we have more information should we ever need it. Would be the same if it's a method somewhere else and it's passed the Bundler::Dependency instance, but then indeed there is no point to move the method.

> Why will you want to override this method?

Because I want some flexibility, for instance I'd think the list might depend on some TruffleRuby option, or depend on the platform, or might want to match gem names with a regexp (e.g., to handle potential forks of those gems, or some related gems which should all be installed as binary).
In other words, a constant like that does not feel like a proper API, a method seems much more flexible.
But I'm OK to override `Bundler::Dependency#default_force_ruby_platform` in TruffleRuby, so that could be the "extension point".

It also feels cleaner to me if TruffleRuby overrides the method and the default is just `false`, because then there are no truffleruby specifics here, except for a comment on this new method to mention it is overwritten by truffleruby (but possibly also by other Ruby implementations which want to override it in the future).

File: bundler/lib/bundler/rubygems_ext.rb [hunk 3]
```diff
@@ -221,6 +221,10 @@ class Platform
     MSWIN64 = Gem::Platform.new("mswin64") unless defined?(MSWIN64)
     MINGW = Gem::Platform.new("x86-mingw32") unless defined?(MINGW)
     X64_MINGW = Gem::Platform.new("x64-mingw32") unless defined?(X64_MINGW)
+
+    if RUBY_ENGINE == "truffleruby" && !defined?(REUSE_AS_BINARY_ON_TRUFFLERUBY)
+      REUSE_BINARY_ON_TRUFFLERUY = %w[libv8 sorbet-static]
```
Comment (eregon) [inline bundler/lib/bundler/rubygems_ext.rb:226]: It's `REUSE_AS_BINARY_ON_TRUFFLERUBY` not `REUSE_BINARY_ON_TRUFFLERUY`.
I'd prefer if we don't depend on that constant name going forward though.

File: bundler/lib/bundler/dependency.rb (hunk 2 above)
Comment (eregon) [reply bundler/lib/bundler/dependency.rb]: > It also feels cleaner to me if TruffleRuby overrides the method and the default is just `false`, because then there are no truffleruby specifics here, except for a comment on this new method to mention it is overwritten by truffleruby (but possibly also by other Ruby implementations which want to override it in the future).

Ah, but that can't be because then existing TruffleRuby releases would just break with latest Bundler, I didn't think about that ealier.
OK, then using the constant on those older TruffleRuby versions seems unavoidable.
Are you OK if newer TruffleRuby releases override this method? Could you add a comment that the method might be overridden by Ruby implementations and e.g. truffleruby?

File: bundler/lib/bundler/rubygems_ext.rb (hunk 3 above)
Comment (eregon) [reply bundler/lib/bundler/rubygems_ext.rb]: (still a typo here, missing a B before the last Y)

File: bundler/lib/bundler/dependency.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply bundler/lib/bundler/dependency.rb]: I think you can send us a patch with whatever comment is needed when/if you need to override the method in the future

File: bundler/lib/bundler/dependency.rb [hunk 4]
```diff
@@ -158,5 +159,13 @@ def specific?
     rescue NoMethodError
       requirement != ">= 0"
     end
+
+    private
+
+    def default_force_ruby_platform
```
Comment (eregon) [inline bundler/lib/bundler/dependency.rb:165]: I'd like to suggest this:
```suggestion
    # Note that TruffleRuby (and possibly other Ruby implementations) might override this method for more flexibility.
    # This default implementation is provided for compatibility with older TruffleRuby releases.
    def default_force_ruby_platform
```

Comment (deivid-rodriguez) [reply bundler/lib/bundler/dependency.rb:165]: As I said above, let's discuss this when/if you decide to override the method in the future. It seems clear to me now what the method is doing.

File: bundler/lib/bundler/man/gemfile.5.ronn [hunk 5]
```diff
@@ -231,6 +231,20 @@ All operations involving groups ([`bundle install`](bundle-install.1.html), `Bun
 `Bundler.require`) behave exactly the same as if any groups not
 matching the current platform were explicitly excluded.
 
+### FORCE_RUBY_PLATFORM
+
+If you always want the pure ruby variant of a gem to be chosen over platform
+specific variants, you can use the `force_ruby_platform` option:
+
+    gem "ffi", :force_ruby_platform => true
```
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply bundler/lib/bundler/man/gemfile.5.ronn:239]: Fair enough, I used our internal style for the documentation, but we should use the most common style.

**Comment by eregon** [comment]
@deivid-rodriguez Thank you for your work on this!
It seems like a great way to implement the behavior for truffleruby, much better than having to change/rely on `Gem.platforms` like I did initially (I didn't realize it'd be enough to change the `force_ruby_platform` on a dependency).

I'll have to check if anything else in RubyGems uses `Gem.platforms` as a way to select gems, I'd guess yes for e.g. `gem install ...`, so probably we still need `match_gem?`, etc for RubyGems, or maybe some newer way piggy-backing on RubyGems's equivalent of the `force_ruby_platform` flag.

File: bundler/lib/bundler/dependency.rb (hunk 4 above)
Comment (eregon) [reply bundler/lib/bundler/dependency.rb:165]: OK. It wasn't clear to me initially on this PR, so I think such a comment would be helpful for anyone else reading this code.

**Comment by deivid-rodriguez** [comment]
> It seems like a great way to implement the behavior for truffleruby, much better than having to change/rely on Gem.platforms like I did initially (I didn't realize it'd be enough to change the force_ruby_platform on a dependency).

I didn't occur to me either until I was working on this feature :)

> I'll have to check if anything else in RubyGems uses Gem.platforms as a way to select gems, I'd guess yes for e.g. gem install ..., so probably we still need match_gem?, etc for RubyGems, or maybe some newer way piggy-backing on RubyGems's equivalent of the force_ruby_platform flag.

Yes, I think `match_gem?`, etc are still needed for RubyGems, but for now it feels nice to be able to simplify the Bundler side.

**Comment by eregon** [comment]
Accidentally closed by https://github.com/rubygems/rubygems/pull/5674

**Comment by deivid-rodriguez** [comment]
Ooops.

**Comment by deivid-rodriguez** [comment]
Alright let's do this!
